File: 19. Remove Nth Node From End of List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        int sz = 0;
        ListNode* cur = head;
        while(cur)
        {
            cur = cur->next;
            ++sz;
        }
        sz -= n;
        ListNode* prev = nullptr;
        cur = head;
        while(sz)
        {
            prev = cur;
            cur = cur->next;
            sz--;
        }
        if(prev)
        {
            prev->next = cur->next;
            delete cur;
        }
        else
        {
            head = cur -> next;
            delete cur;
        }
        return head;
    }
};

File: 19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        fast = head
        slow = head
        prev = None
        ret = head

        while n != 0:
            fast = fast.next
            n -= 1
        if not fast:
            return head.next
        
        while fast != None:
            fast = fast.next
            prev = slow
            slow = slow.next
        
        if prev:
            prev.next = slow.next
            return ret
        else:
            return None
